File: engine/src/com/magma/engine/stages/StageSwitcher.java
package com.magma.engine.stages;

import com.badlogic.gdx.Screen;
import com.badlogic.gdx.scenes.scene2d.Stage;
import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.Disposable;
import com.magma.engine.MagmaGame;
import com.magma.engine.debug.modules.StageModuleListener;

//TODO: Add stage transitions
public class StageSwitcher implements Screen, Disposable, StageModuleListener {

    private static Array<StageSwitchListener> listeners;
	private static StageSwitcher instance;
    private static GameStage active;

    private StageSwitcher() {
        listeners = new Array<StageSwitchListener>();
        MagmaGame.disposeOnExit(this);
    }

    public static StageSwitcher getInstance() {
        if (instance == null){
            instance = new StageSwitcher();
        }
        return instance;
    }

    public static void open(GameStage stage) {
        StageSwitcher.active = stage;

        // make the stage a listener if it has the correct interface
        if (stage instanceof StageSwitchListener) {
            listeners.add((StageSwitchListener) stage);
        }

        // notify all the listeners
        for (StageSwitchListener s : listeners) {
            s.stageOpened(stage);
        }
    }

    public static void addListener(StageSwitchListener listener) {
        listeners.add(listener);
    }

    public static void act(float delta) {
        active.act(delta);
    }

    @Override
    public void render(float delta) {
        if (active != null) {
            active.getViewport().apply();
            active.draw();
        }
    }

    @Override
    public void dispose() {
        if (active != null) {
            active.dispose();
        }
        listeners = null;
        instance = null;
    }

    @Override
    public void resize(int width, int height) {
        if (active != null) {
            active.getViewports().resize(width, height);
            Stage ui = active.getUIStage();
            ui.getViewport().update(width, height);
            ui.getCamera().update(true);
        }
        for (StageSwitchListener listener : listeners){
            listener.resize(width, height);
        }
    }

    @Override
    public void pause() {
    }

    @Override
    public void resume() {
    }

    @Override
    public void show() {
    }

    @Override
    public void hide() {
    }

    @Override
    public void unloadStage() {

    }
}

File: engine/src/com/magma/engine/stages/StageSwitcher.kt
package com.magma.engine.stages

import com.badlogic.gdx.Screen
import com.badlogic.gdx.utils.Array
import com.badlogic.gdx.utils.Disposable
import com.magma.engine.MagmaGame.Companion.disposeOnExit
import com.magma.engine.debug.modules.StageModuleListener

//TODO: Add stage transitions
object StageSwitcher : Screen, Disposable, StageModuleListener {
    private val listeners: Array<StageSwitchListener> = Array()
    private lateinit var active: GameStage
    init {
        disposeOnExit(this)
    }

    override fun render(delta: Float) {
        active.viewport.apply()
        active.draw()
    }

    override fun dispose() {
        active.dispose()
        listeners.clear()
    }

    override fun resize(width: Int, height: Int) {
        ViewportContext.resize(width, height)
        val ui = active.uiStage
        ui.viewport.update(width, height)
        ui.camera.update(true)
        for (listener in listeners) {
            listener.stageResized(width, height)
        }
    }

    fun open(stage: GameStage) {
        active = stage

        // make the stage a listener if it has the correct interface
        if (stage is StageSwitchListener) {
            listeners.add(stage as StageSwitchListener)
        }

        // notify all the listeners
        for (s in listeners) {
            s.stageOpened(stage)
        }
    }

    fun addListener(listener: StageSwitchListener) {
        listeners.add(listener)
    }

    fun act(delta: Float) {
        active.act(delta)
    }

    override fun pause() {}
    override fun resume() {}
    override fun show() {}
    override fun hide() {}
    override fun unloadStage() {}
}